### PR TITLE
Return UNESCAPED_CHARS immediately

### PR DESCRIPTION
**Pull request by jkeiser**
Doing this gives a very small (but real) performance improvement in stage 1 on Kady Lake (Haswell+), and also reduces the amount of indirection, which will improve readability/grokkability.

Note: I attempted to do the same thing with the utf8 check, but that had a negative effect. Presumably this is because the UTF-8 error mask is in SIMD, and it's a lot more expensive to pull data from SIMD to scalar world with operations like test and movmskb. unescaped_chars_error was already scalar and didn't have that problem.

### CODE REVIEW

**Comment by jkeiser** [comment]
Also interesting: this is one of the cases where `unlikely` was absolutely necessary. I think the rule is probably *any* work in either the if condition or the conditional block has impact unless you use `unlikely`. Passthrough if statements (like `if (error) { return error; }`) do not require `likely`.

**Comment by lemire** [comment]
@jkeiser 

I think we need to be prudent when it comes to replacing branchless code with branchy code because benchmarks can easily favor branchy code in benchmarks for misleading reasons.  See my post [Benchmarking is hard: processors learn to predict branches](https://lemire.me/blog/2019/10/16/benchmarking-is-hard-processors-learn-to-predict-branches/).

Do we have a rationale for why we think that your code should be more performant?


In at least one instance, I see a performance degradation. On a skylake processor with GCC 8.3.

Before:
```
jsonexamples/twitter.json
stage 1 instructions:    1711881 cycles:     481401 (48.86 %) ins/cycles: 3.56 mis. branches:        982 (cycles/mis.branch 490.06) cache accesses:      29169 (failure          0)
 stage 1 runs at 0.76 cycles per input byte.
```



After:
```
jsonexamples/twitter.json
stage 1 instructions:    1728989 cycles:     489486 (49.02 %) ins/cycles: 3.53 mis. branches:       1234 (cycles/mis.branch 396.46) cache accesses:      29230 (failure          0)
 stage 1 runs at 0.78 cycles per input byte.
```



Here is the full output:


Before:

Trial 1
```
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1197 cycles:        740 (0.08 %) ins/cycles: 1.62 mis. branches:          3 (cycles/mis.branch 239.76) cache accesses:      29160 (failure          0)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    1711881 cycles:     480798 (48.98 %) ins/cycles: 3.56 mis. branches:        989 (cycles/mis.branch 486.07) cache accesses:      29160 (failure          2)
 stage 1 runs at 0.76 cycles per input byte.
stage 2 instructions:    1490176 cycles:     500122 (50.95 %) ins/cycles: 2.98 mis. branches:       1522  (cycles/mis.branch 328.48)  cache accesses:      54494 (failure          5)
 stage 2 runs at 0.79 cycles per input byte and 9.05 cycles per structural character.
 all stages: 1.55 cycles per input byte.
Estimated average frequency: 3.715 GHz.
Min:  0.000264275 bytes read: 631515 Gigabytes/second: 2.38961
```

Trial 2
```
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1197 cycles:        725 (0.07 %) ins/cycles: 1.65 mis. branches:          2 (cycles/mis.branch 289.09) cache accesses:      29112 (failure          0)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    1711881 cycles:     481567 (48.88 %) ins/cycles: 3.55 mis. branches:        992 (cycles/mis.branch 485.11) cache accesses:      29112 (failure          0)
 stage 1 runs at 0.76 cycles per input byte.
stage 2 instructions:    1490176 cycles:     502892 (51.05 %) ins/cycles: 2.96 mis. branches:       1529  (cycles/mis.branch 328.74)  cache accesses:      54562 (failure          0)
 stage 2 runs at 0.80 cycles per input byte and 9.10 cycles per structural character.
 all stages: 1.56 cycles per input byte.
Estimated average frequency: 3.714 GHz.
Min:  0.000265279 bytes read: 631515 Gigabytes/second: 2.38057
```


After:

Trial 1
```
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1197 cycles:        840 (0.08 %) ins/cycles: 1.42 mis. branches:          4 (cycles/mis.branch 209.49) cache accesses:      29254 (failure          0)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    1728989 cycles:     493457 (49.19 %) ins/cycles: 3.50 mis. branches:       1231 (cycles/mis.branch 400.84) cache accesses:      29254 (failure          0)
 stage 1 runs at 0.78 cycles per input byte.
stage 2 instructions:    1490176 cycles:     508791 (50.72 %) ins/cycles: 2.93 mis. branches:       1780  (cycles/mis.branch 285.82)  cache accesses:      54636 (failure          0)
 stage 2 runs at 0.81 cycles per input byte and 9.21 cycles per structural character.
 all stages: 1.59 cycles per input byte.
Estimated average frequency: 3.728 GHz.
Min:  0.000269056 bytes read: 631515 Gigabytes/second: 2.34715
```

Trial 2
```
number of iterations 1000 
number of bytes 631515 number of structural chars 55264 ratio 0.088
mem alloc instructions:       1197 cycles:        778 (0.08 %) ins/cycles: 1.54 mis. branches:          2 (cycles/mis.branch 312.16) cache accesses:      29231 (failure          0)
 mem alloc runs at 0.00 cycles per input byte.
stage 1 instructions:    1728989 cycles:     489177 (48.91 %) ins/cycles: 3.53 mis. branches:       1232 (cycles/mis.branch 396.84) cache accesses:      29231 (failure          0)
 stage 1 runs at 0.77 cycles per input byte.
stage 2 instructions:    1490176 cycles:     510193 (51.01 %) ins/cycles: 2.92 mis. branches:       1780  (cycles/mis.branch 286.59)  cache accesses:      54683 (failure          1)
 stage 2 runs at 0.81 cycles per input byte and 9.23 cycles per structural character.
 all stages: 1.58 cycles per input byte.
Estimated average frequency: 3.730 GHz.
Min:  0.000268172 bytes read: 631515 Gigabytes/second: 2.35489
```


So I suggest we go forward by examining the benchmarks more closely. What are your numbers telling you?

**Comment by jkeiser** [comment]
I'll send the bench info over when I get home later.

I tried this because I noticed an improvement when I changed an error if statement to. My reasoning is that branch prediction of branches we never take (in the normal course) ought to be near enough to perfect as to be noise. As far as I understand, it's not branches that are the problem, it's mispredicts, right?

I'm ignorant of how compilers optimize `unlikely`, but my expectation is it lets the compiler know it can safely move the if block out of the code cache line so more likely instructions can be decoded earlier, and probably lets it schedule the if condition later than usual to allow more data critical statements to execute earlier.

Using a return means that we break a data dependency (the error variable) between one iteration of the loop and the next, allowing them to run somewhat more out of order, too.

I was surprised when I discovered this, but I think it makes sense. It'll be interesting to look what's going on in Sandy bridge!

I'll mark it engineering research as well, since it sounds like it's not a clear improvement across all architectures. It's really not a large enough improvement that we'd want to drop everything to get it in, anyway :)

**Comment by lemire** [comment]
> Using a return instead of error |= more_errors also means that we break a data dependency between one iteration of the loop and the next, allowing them to run somewhat more out of order, too.

You may be right, but I am skeptical. Can you elaborate?

**Comment by lemire** [comment]
> I'm ignorant of how compilers optimize unlikely, but my layman's expectation is unlikely lets the compiler know it can safely move the if block out of the code cache line so more likely instructions can be decoded earlier. It also probably lets it schedule the if condition instructions later than usual to allow more critical statements to execute earlier.

I think that what it does, often, is just to move out the instructions corresponding to the unlikely choice entirely off the main code sequence so that no jump is typically needed and you have a very compact code sequence (without the unlikely instructions occupying cache memory or requiring decoding). Of course, who knows what else it might do...?

**Comment by lemire** [comment]
@jkeiser  Note... I am still hoping we can mover forward, I just want us to discuss the performance effects with actual numbers.

To be clear, I am not proposing to block this PR. I just think we need to document and discuss, if only to leave a trace.

**Comment by jkeiser** [comment]
> > Using a return instead of error |= more_errors also means that we break a data dependency between one iteration of the loop and the next, allowing them to run somewhat more out of order, too.
> You may be right, but I am skeptical. Can you elaborate?

The basic idea is, if you have `int error` and a loop (possibly partially unrolled) that does `error |= <something>` on every input, the `|=` becomes a logjam: each <something> has to be fully computed and `|`'d into the `error` register before the next iteration of the loop can be computed.

On the other hand, if you use `if (unlikely(<something>)) { return false; }`, each loop iteration can be processed independently, and the only dependency is really on the branch predictor, which will assume `false` every time and be right (since we're optiimizing for no-error cases).

Let me know if that makes sense to you. While I believe all of what I just said, I don't trust my model of microarchitecture as much as yours: I'm partly making and testing theories to be knocked down, so I can learn something :)

> I am still hoping we can move forward, I just want us to discuss the performance effects with actual numbers.

I actually only want to do this if it actually improves performance across the board. While I have arguments, it's basically an attempt to see if the theory pans out in reality. It's very slightly nicer, but there are very few times I would want to trade readability for performance in this project. I want to find ways to make it readable *without* performance loss :)

**Comment by lemire** [comment]
@jkeiser 

> Let me know if that makes sense to you. 

It does not, but it does not mean that you are wrong. Let me know if you think I am making sense.

Let us model it out.

```
  do work A
  error |= result from A
  do work B
  error |= result from B
```

Means that you can do...

```
  do work A AND do work B
  error |= result from A
  error |= result from B
```

So yeah, you have a data dependency, I guess, so you may not be able to do the last two lines at the same time... but unless the work you do outside of these two lines is super cheap, they are not going to be a bottleneck. A bitwise OR is super cheap.

However, should it ever become a bottleneck, you can double it up...

```
  do work A AND do work B
  errorA |= result from A AND errorB |= result from B
```

Then later do `error = errorA | errorB`.

Let us compare with the if-then approach...

```
  do work A
  if( result from A) error
  do work B
  if (result from B) error
```

But this can be reordered...

```
  do work A AND do work B
  if( result from A) error
  if (result from B) error
```

I doubt you can do two branches in the same cycle... I think that's not possible on any platform.

So it is the same thing... except that the branch is slightly more expensive than a bitwise OR. It is more expensive in the sense that you have more instructions, though they may fuse so you could end up with the same muops count... but they may harm branch prediction (this is speculative on my part).


A case that can be made, I think, is that the performance is nearly the same (that I believe) that the branching may help with error recovery. What I don't believe, at this point, is that we should expect better performance by the introduction of branching.

**Comment by jkeiser** [comment]
Hmm, I think the point there is that since error is basically a terminal (nothing depends on it), even if two iterations of the loop depend on each other and get bogged down they can't block anything *else* from running. I can buy that.

If the if statements are the same size, the only benefits of moving errors to an if statement would be:

- less register pressure if that's an issue (though it's possible everything will be just as fast if error is in memory since it's a non blocking operation).
- allowing the use of `unlikely` to move hefty conditional blocks out (you could just as easily keep the block small by moving the work to a non-inline function).
- *Maybe* the use of unlikely to let the compiler shift the if *condition* to the most unobtrusive place it can

In short, I think I'm convinced it's not necessary to do this :)

By keeping error separate there is also the possibility that we could do some bit magic to delay some of the actual error detection ... like the fact that you could `|` all the bytes together and test the high bit once at the end to detect non ASCII (that specific algorithm won't help is but things like that might).

**Comment by lemire** [comment]
Relevant blog post: [Adding a (predictable) branch to existing code can increase branch mispredictions ](https://lemire.me/blog/2019/11/06/adding-a-predictable-branch-to-existing-code-can-increase-branch-mispredictions/).
